### PR TITLE
raftstore: return more region info

### PR DESCRIPTION
**Pull request by BusyJay**
### What problem does this PR solve?

Issue Number: close #xxx 

Problem Summary:

### What is changed and how it works?

Proposal: [xxx](url) 

What's Changed:

### Related changes

- PR to update `pingcap/docs`/`pingcap/docs-cn`:
- PR to update `pingcap/tidb-ansible`:
- Need to cherry-pick to the release branch

### Check List 

Tests 

- Unit test
- Integration test
- Manual test (add detailed scripts or steps below)
- No code

Side effects

- Performance regression
    - Consumes more CPU
    - Consumes more MEM
- Breaking backward compatibility

### Release note 
- No release note

### CODE REVIEW

File: components/raftstore/src/store/fsm/peer.rs [hunk 1]
```diff
@@ -3047,17 +3047,45 @@ impl<'a, T: Transport, C: PdClient> PeerFsmDelegate<'a, T, C> {
...
+            let res = if self.ctx.cfg.right_derive_when_split {
+                ranges.next_back()
+            } else {
+                ranges.next()
+            };
+            if let Some((_, id)) = res {
+                let r = &meta.regions[id];
+                collect_cnt -= 1;
+                // For example, A is split into B, A, and then B is split into C, B.
+                if r.get_region_epoch().version >= max_version {
```
Comment (tabokie) [inline components/raftstore/src/store/fsm/peer.rs:3076]:
```suggestion
                if r.get_region_epoch().version > max_version {
```

File: components/raftstore/src/store/fsm/peer.rs [hunk 2]
```diff
@@ -3047,17 +3047,45 @@ impl<'a, T: Transport, C: PdClient> PeerFsmDelegate<'a, T, C> {
...
+                    // It doesn't matter if it's a false positive, as it's limited by MAX_REGIONS_IN_ERROR.
+                    collect_cnt += r.get_region_epoch().version - max_version;
+                    max_version = r.get_region_epoch().version;
```
Comment (tabokie) [inline components/raftstore/src/store/fsm/peer.rs:3079]: Do we need to recheck `requested_version >= max_version` here?

Comment (BusyJay) [reply components/raftstore/src/store/fsm/peer.rs:3080]: After L3053, `requested_version < max_version` is true. And `max_version` only goes larger, hence `requested_version` is always less then `max_version`.

**Review by hicqu — APPROVED** [review]

File: components/pd_client/src/client.rs [hunk 3]
```diff
@@ -420,13 +424,27 @@ impl PdClient for RpcClient {
             let (tx, rx) = mpsc::unbounded();
             tx.unbounded_send(req)
                 .unwrap_or_else(|e| panic!("send request to unbounded channel failed {:?}", e));
+            let pending_heartbeat = Arc::new(AtomicU64::new(1));
             inner.hb_sender = Either::Right(tx);
-            Box::new(
+            inner.pending_heartbeat = pending_heartbeat.clone();
```
Comment (overvenus) [inline components/pd_client/src/client.rs:429]: Please add a comment about the reset.

**Review by NingLin-P — APPROVED** [review]

File: components/pd_client/src/client.rs [hunk 4]
```diff
@@ -407,6 +409,8 @@ impl PdClient for RpcClient {
         let executor = |client: &RwLock<Inner>, req: pdpb::RegionHeartbeatRequest| {
             let mut inner = client.wl();
             if let Either::Right(ref sender) = inner.hb_sender {
+                let last = inner.pending_heartbeat.fetch_add(1, Ordering::Relaxed);
+                PD_PENDING_HEARTBEAT_GAUGE.set(last as i64 + 1);
```
Comment (overvenus) [inline components/pd_client/src/client.rs:413]: Seem it is unnecessary, receiver updates the metric too.

**Review by Connor1996 — APPROVED** [review]
LGTM

**Comment by BusyJay** [comment]
All changes are merged into master.

**Pull request closed without merging**